fix(linter)!: override plugins array when passed in config file

camchenry opened the pull request:
- fixes https://github.com/oxc-project/oxc/issues/6896

When the `plugins` config property is specified, it will overwrite the default plugins array. This allows the plugins list to be easily customized and allows for disabling default plugins at the same time as enabling non-default ones.

- example: `{ "plugins": [] }` will enable no plugins.
- example: `{ }` will enable default plugins.
- example: `{ "plugins": ["typescript", "import"] }` will enable only the import and typescript plugins.

Sysix commented:
Thanks on working on this. ❤️ With the new [overrides](https://github.com/oxc-project/oxc/pull/6974/files) property there is a new discussion to make :) 

```
{
  plugins: ["typescript"],
  overrides: [
    {
        files: ["*.tsx"],
        plugins: ["react"]
    }
  ]
}
```

with this config I expect to enable ts and react plugin for `.tsx` files. Maybe someone other expect it differently :)

Sysix commented:
That is a also a topic 👍 Let not block this PR, perrsonally im fine at the moment.

> if i specify an empty array. but have a rule in the rules list that uses a plugin does it get reported as an unknown rule?

existing rules in the project should always be known and can be enabled/disabled in the rules block.
plugins + categories enables the base rules, rules block should adjust the last needs.

That what I expect as a user, don't know what the core team says about it :)

camchenry commented:
> Thanks for working on this. ❤️ With the new [overrides](https://github.com/oxc-project/oxc/pull/6974/files) property there is a new discussion to make :)
> 
> ```
> {
>   plugins: ["typescript"],
>   overrides: [
>     {
>         files: ["*.tsx"],
>         plugins: ["react"]
>     }
>   ]
> }
> ```
> 
> with this config I expect to enable ts and react plugin for `.tsx` files. Maybe someone other expect it differently :)

Good point, I forgot to consider the `overrides` property here. I think that making the plugins array append when inside of overrides feels the most intuitive to me? I can update the tests and behavior accordingly.

> if i specify an empty array. but have a rule in the rules list that uses a plugin does it get reported as an unknown rule?

Yes, I think this should report as an unknown/disabled rule warning. I can imagine a case where you remove a plugin that is no longer needed and want to be warned of config that is no longer useful. Or, like in the case above with overrides, you need to explicitly enable the plugin to use the rule.

Sysix commented:
> I think that making the plugins array append when inside of overrides feels the most intuitive to me? 

Feels for me too, but I do not know how it is currently with `categories`. 
When `overrides` **_really overrides_** and not append. I think my expection is wrong and we should keep it consistent.

camchenry commented:
I checked the behavior with the legacy `.eslintrc` setup in ESLint, and it this config will append so that `.ts` files have the react and typescript plugins enabled:

```json
{
  "plugins": ["@typescript-eslint"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["react"]
    }
  ]
}
```

So, I think this is the way to go for compatibility. And specifying an empty array here does not disable all of the plugins.

Boshen commented:
### Merge activity

* **Nov 16, 12:41 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
fix(linter)!: override plugins array when passed in config file (#7303)

- fixes https://github.com/oxc-project/oxc/issues/6896

When the `plugins` config property is specified, it will overwrite the default plugins array. This allows the plugins list to be easily customized and allows for disabling default plugins at the same time as enabling non-default ones.

- example: `{ "plugins": [] }` will enable no plugins.
- example: `{ }` will enable default plugins.
- example: `{ "plugins": ["typescript", "import"] }` will enable only the import and typescript plugins.
